internal/controllers/machineset/machineset_controller.go

```diff
@@ -966,6 +994,146 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubernetesVersionPreflightCheck(ctx context.Context, cpVersion string, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// If the version is not defined on the MachineSet then we cannot check for the version skew policies.
+	// Return early.
+	if ms.Spec.Template.Spec.Version == nil {
+		return ctrl.Result{}, nil
+	}
+	msVersion := *ms.Spec.Template.Spec.Version
+
+	cpSemver, err := semver.ParseTolerant(cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse ControlPlane version %s", cpVersion)
+	}
+	msSemver, err := semver.ParseTolerant(msVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform kubernetes version skew pre-flight check: failed to parse MachineSet version %s", msVersion)
+	}
+
+	// Check the Kubernetes version skew policy.
+	// => MS minor version cannot be greater than the Control Plane minor version.
+	// => MS minor version cannot be older than 2 minor version of Control Plane.
+	// Kubernetes skew policy: https://kubernetes.io/releases/version-skew-policy/#kubelet
+	floorVersion := semver.Version{
+		Major: cpSemver.Major,
+		Minor: cpSemver.Minor - 2,
+		Patch: 0,
+	}
+	if msSemver.GT(cpSemver) || msSemver.LT(floorVersion) {
+		log.Info(fmt.Sprintf("Pre-flight check failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy", msVersion, cpVersion))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubeadmVersionPreflightCheck(ctx context.Context, cpVersion string, ms *clusterv1.MachineSet) ctrl.Result {
+	log := ctrl.LoggerFrom(ctx)
+	// If the version is not defined on the MachineSet then we cannot check for the version skew policies.
+	// Return early.
+	if ms.Spec.Template.Spec.Version == nil {
+		return ctrl.Result{}
+	}
+	msVersion := *ms.Spec.Template.Spec.Version
+
+	// If using kubeadm bootstrap provider, check the kubeadm version skew policy.
+	// => MS version should match the Control Plane version.
+	// kubeadm skew policy: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#kubeadm-s-skew-against-kubeadm
+	kubeadmBootstrapProviderUsed := ms.Spec.Template.Spec.Bootstrap.ConfigRef != nil &&
+		ms.Spec.Template.Spec.Bootstrap.ConfigRef.Kind == "KubeadmConfigTemplate"
+	if kubeadmBootstrapProviderUsed {
+		if cpVersion != msVersion {
+			log.Info(fmt.Sprintf("Pre-flight check failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to kubeadm version skew policy: kubeadm bootstrap provider only supports joining with the same version as the control plane", msVersion, cpVersion))
```

Comment:
Hm, shouldn't this allow users to create worker nodes with an older version that's still in the skew policy?

Reply:
The kubeadm skew policy is pretty strict:

> If new nodes are joined to the cluster, the kubeadm binary used for kubeadm join must match the last version of kubeadm used to either create the cluster with kubeadm init or to upgrade the same node with kubeadm upgrade.
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#version-skew-policy

Reply:
This skew policy is for the kubeadm binary, not for the kubelet version which joins the cluster (kubeadm even tests joining with an older kubelet binary). 

So the preflight check would make the assumption, that the version set at the machineset matches kubeadm version used if I'm reading it correctly? 🤔

Reply:
@chrischdi The version set here determines the VM image (or the appropriate infra specific thing) used and the assumption is that VM image uses the matching kubeadm version binary (which I believe is the is the case for the providers).

@vincepri This preflight check is only performed if using the kubeadm bootstrap provider. The preflight check will only fail when trying to create new Machines (scale down still works) using the old version because `kubeadm join` could fail as the kubeadm version skew is violated. 
Along with that, the user can also just skip the kubeadm preflight check if the user has verified that new workers running an old version can join the control plane.

Reply:
The problem is that joining new nodes with an older kubeadm version is not supported by kubeadm upstream. It mostly works but there is no test coverage for it.

An example can be seen here: https://github.com/kubernetes-sigs/cluster-api/issues/7533#issuecomment-1319222151. The old kubeadm version wasn't able to parse the ClusterConfiguration written by the newer kubeadm version: (or in the Cluster API case, KCP)
> error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to decode cluster configuration data: no kind "ClusterConfiguration" is registered for version "kubeadm.k8s.io/v1beta3" in scheme "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/scheme/scheme.go:31"
But this is just an example we're aware of, as there is no testing on the kubeadm side there simply is no official support for this.

Our idea was to have a preflight check to ensure we're acting within the official support ranges of kubeadm & Kubernetes. For folks who are sure that other combinations work as well (I assume because they tested it) we wanted to provide a way to opt-out.

There are a few cases where it's better to not act if we're outside of the support range and thus can't tell if the Machine has a chance to come up:
* MHC: It's better to keep an (at least momentarily) unhealty Machine vs deleting it and then the creation/join fails
* If autoscaler scales up a MD/MS with an older version and the Machine creations/joins continously fail we can have a lot of Machines around which have no chance for success but can become pretty expensive